### PR TITLE
Add ROS 2 command line arguments design document

### PR DESCRIPTION
**Pull request by hidmic**
Follow-up PR after the discussions that took place in #242.

### CODE REVIEW

**Comment by hidmic** [comment]
I think we can (and probably should) augment this document to encompass all ROS 2 nodes command line features. Also, I wonder now that we're introducing options, if we should move some features away from the double leading underscore syntax e.g. `__log_level:=...` to `--log-level ...`. It's a cosmetic change though.

File: articles/150_ros_command_line.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+As it was the case in ROS 1, ROS 2 nodes allow configuration via command line arguments to a certain degree.
+However, a larger set of configuration options, an ambiguity in remapping rules and parameter assignment syntax (as a result of the leading underscore name convention for hidden resources), a one-to-many relationship between executables and nodes, among others, add up complexity to an otherwise simple interface.
```
Comment (wjwwood) [inline articles/150_ros_command_line.md]: I'd reword this to have the conclusion then followed by the reasons (just because it's hard to see how the enumerated items are related which reading).

Something like this (might need more polish, but just an example of flipping the order of the sentence):

> However, in ROS 2 the complexity of these command line arguments have increased due to: a larger set of configuration options, an ambiguity in remapping rules and parameter assignment syntax (as a result of the leading underscore name convention for hidden resources), a one-to-many relationship between executables and nodes, etc.

File: articles/150_ros_command_line.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+Increasingly precise addressing mechanisms, as well as leading double underscores (`__`) in positional arguments, solve part of the problem.
+In ROS 2, however, command line options are introduced too as these are:
```
Comment (wjwwood) [inline articles/150_ros_command_line.md]: For me "command line options" includes the "positional" `from:=to` style arguments. Something like `--opt` or `-o` is what I'd call "flags". I'm not sure that's the right wording, but as it's written it's a confusing. Maybe something like:

> In ROS 2, however, additional command line options have been introduced in order to be:

File: articles/150_ros_command_line.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+- Explicit and less error prone.
+- Suitable for tab completion.
+
+Unfortunately, additional devices must be put in place as these coexist with user-defined options -- part of the rationale behind ROS 1 command line design.
+
+## Namespacing
+
+To prevent ROS specific command line options from interacting with user-defined ones, the former are scoped using the `--ros-args` option and a trailing double dash token (`--`):
```
Comment (wjwwood) [inline articles/150_ros_command_line.md]: Personally, I would use `interfering` or `colliding` rather than `interacting`.

File: articles/150_ros_command_line.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+```sh
+ros2 run some_package my_node [<user-defined-pre-arg0>...] \
+  --ros-args <ros-specific-arg-0> [<ros-specific-arg-1>...] -- <user-defined-arg-1> [<user-defined-arg-2>...]
+```
+
+If no user defined arguments are provided after ROS specific arguments are, the double dash token (`--`) may be elided:
+
+```sh
+ros2 run some_package my_node [<user-defined-arg-0>...] --ros-args <ros-specific-arg-0> [<ros-specific-arg-1>...]
+```
```
Comment (wjwwood) [inline articles/150_ros_command_line.md]: One more thing to clarify, can you have more than one instance of `--ros-args ... --`? I would say we should so that multiple sources of ROS options can just append them to the command line arguments without having to worry if there was a previous instance or not.

File: articles/150_ros_command_line.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+Other, alternative designs were under discussion.
+
+### Additional operators
+
+Stop using the same `:=` operator for parameter assignments and name remapping rules and introduce additional operators e.g. `:=` for parameter assignment and `~=` for name remapping. This keeps the command line verbosity at a minimum, but is error prone.
+
+### Full name addressing
+
+Rely on full name addressing to disambiguate operator significance e.g. `rosparam://this:=that` would mean to assign a `that` string value to parameter `this` while `rostopic://this:=that` would mean to remap topic `this` to `that`. This signficantly increases the command line verbosity.
```
Comment (wjwwood) [inline articles/150_ros_command_line.md]: Wouldn't it need to be `rosremap://` or something? Right now remaps apply to topics and services.

File: articles/150_ros_command_line.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+### Prefixed option names
+
+Remove the need for double dash tokens (`--`), conventionally used to signify the end of CLI options for a command, by adding the `--ros-` prefix to all ROS specific command line options e.g. `--ros-remap`, `--ros-param`, etc. This somewhat increases the command line verbosity.
```
Comment (wjwwood) [inline articles/150_ros_command_line.md]: I would also say that this one makes the extraction of the ROS arguments slightly more difficult because you need to know about all ros options ahead of time, whereas the currently proposal will always be removable with just two or three rules to consider.

File: articles/150_ros_command_line.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+ros2 run some_package some_node --ros-args -p int_param:=2
+```
+
+## Alternative designs
+
+Other, alternative designs were under discussion.
+
+### Additional operators
+
+Stop using the same `:=` operator for parameter assignments and name remapping rules and introduce additional operators e.g. `:=` for parameter assignment and `~=` for name remapping. This keeps the command line verbosity at a minimum, but is error prone.
```
Comment (jacobperron) [inline articles/150_ros_command_line.md]: nit: one sentence per line

It might be worth clarifying that this option avoids the need for options at all (ie. no need for `--ros-args`).

File: articles/150_ros_command_line.md (hunk 2 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: Yeah, it's poorly written. I almost rewrote the entire section in beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md (hunk 3 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: I like `colliding`. See beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md (hunk 4 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: That's a very good point. See beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md (hunk 5 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: To force a remap for all matching interfaces regardless of their type? Sure, yeah. See beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md (hunk 6 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: True that. See beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md (hunk 7 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: Done in beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md (hunk 1 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: Reworded in beaf7bec3dfc45db907f8c89e30801dcfdc3eba1.

File: articles/150_ros_command_line.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+```sh
+ros2 run some_package some_node --ros-args -r foo:=bar
+```
+
+### Parameter assignment statements
+
+Parameter assignment may be achieved using the `--param`/`-p` option.
+This option takes a single `name:=value` assignment statement, where `value` is in YAML format.
+
+As an example, to assign an integer value of `2` to an `int_param` upon running `some_node`, one may execute:
```
Comment (dirk-thomas) [inline articles/150_ros_command_line.md]: I am not sure if `integer` is a good example here. That also opens the question what makes this a parameter of type since there needs to be some magic to convert the type solely based on its value which has problems on its own.

Therefore I would suggest to simply use a string value here.

File: articles/150_ros_command_line.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+title: ROS Command Line
+permalink: articles/ros_command_line.html
+abstract:
+  This article describes ROS 2 nodes command line capabilities and its syntax.
+author: '[Michel Hidalgo](https://github.com/hidmic)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
```
Comment (dirk-thomas) [inline articles/150_ros_command_line.md]: Maybe add a short section about the implementation. Basically where the extraction from `argc` / `argv` happens and that all remaining arguments are still available for the user code to process.

File: articles/150_ros_command_line.md (hunk 8 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: Hmm, I agree with using a string value to simplify the example but I also think that parameter value syntax cannot be left out of a command line specification. Here, just like when [parameter files are parsed in `rcl`](https://github.com/ros2/rclcpp/blob/8c48dbede7de2bc7e5199f340b6bfd33ffd6a784/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L137) and when [`launch` substitutions occur in `Node` parameters](https://github.com/ros2/launch_ros/blob/master/launch_ros/launch_ros/utilities/evaluate_parameters.py#L70), we delegate on YAML format rules for type inference based on values. Whether that's a good call or not is likely worth a separate discussion.

File: articles/150_ros_command_line.md (hunk 9 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: That makes sense. It also raises an interesting question. Assuming extraction happens in `rcl`, should we support arguments specific to upper layers e.g. `rclcpp` or `rclpy`? Those would have to be available unparsed, separate from user defined arguments.

Comment (wjwwood) [reply articles/150_ros_command_line.md]: We could, but they could live in between the `--ros-args` and `--`, and `rcl` could just ignore them and leave them in a "left over args" output. Then `rclcpp` and/or `rclpy` can use those arguments or make warnings/errors because there are unknown parameters left over.

**Review by wjwwood — APPROVED** [review]

**Comment by hidmic** [comment]
@dirk-thomas PTAL

File: articles/150_ros_command_line.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
+---
+layout: default
+title: ROS Command Line
+permalink: articles/ros_command_line.html
```
Comment (dirk-thomas) [inline articles/150_ros_command_line.md]: Change to: `ros_command_line_arguments`

File: articles/150_ros_command_line.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+## Features
+
+### Namespacing
+
+To prevent ROS specific command line flags from colliding with user-defined ones, the former are scoped using the `--ros-args` flag and a trailing double dash token (`--`):
+
+```sh
+ros2 run some_package some_node [<user-defined-arg0>...] \
+  --ros-args <ros-specific-arg-0> [<ros-specific-arg-1>...] -- <user-defined-arg-1> [<user-defined-arg-2>...]
```
Comment (dirk-thomas) [inline articles/150_ros_command_line.md]: Does `--ros-args` need to be `nargs='+'`? Can't it be defined as `nargs='*'` allowing no ROS specific arguments to follow?

File: articles/150_ros_command_line.md (hunk 10 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: Changed in 4c1b29833c98640308a183a28d9a9b35e8ed1ca5.

File: articles/150_ros_command_line.md (hunk 11 above)
Comment (hidmic) [reply articles/150_ros_command_line.md]: Hmm, what use case do you have in mind?

Comment (dirk-thomas) [reply articles/150_ros_command_line.md]: Just common logic to build these arguments. If at least one is required the logic needs to double check / special case that when generating the command line.

It is the same for any other argument taking multiple parameters in command line tools, e.g. `colcol build --cmake-args`, `ros2 security generate_artifacts --node-names`.

Comment (wjwwood) [reply articles/150_ros_command_line.md]: I don't understand what you mean.

Comment (dirk-thomas) [reply articles/150_ros_command_line.md]: `ros2 run some_package some_node --ros-args` should be a valid invocation. The current design requires at least one argument (`<ros-specific-arg-0>`) to be present.

Comment (wjwwood) [reply articles/150_ros_command_line.md]: I understand now, yeah, I think it should be fine to not actually have any ros arguments follow `--ros-args`, so `--ros-args --` would also be ok.

**Review by dirk-thomas — APPROVED** [review]

**Comment by hidmic** [comment]
@dirk-thomas @wjwwood PTAL at https://github.com/ros2/design/pull/245/commits/c079dc73cc07f5f2e91b2406fc2a5afccb66fd0f so we can call it done.

**Review by dirk-thomas — APPROVED** [review]
lgtm

**Comment by hidmic** [comment]
Alright, going in!

**Review by wjwwood — APPROVED** [review]
late lgtm too :)
